write_xar: move libxml2 behind an abstraction layer

DHowett opened the pull request:
This commit prepares the XAR writer for another XML writing backend.

Almost everything in this changeset leaves the code identical to how
it started, except for a new layer of indirection between the xar writer
and the XML writer.

The things that are not one-to-one renames include:

- The removal of `UTF8Toisolat1` for the purposes of validating UTF-8
  - The writer code made a copy of every filename for the purposes of checking
    whether it was Latin-1 stored as UTF-8. In xar, Non-Latin-1 gets stored
    Base64-encoded.
  - I've replaced this use because (1) it was inefficient and (2) `UTF8Toisolat1`
    is a `libxml2` export.
  - The new function has slightly different results than the one it is replacing
    for invalid UTF-8. Namely, it treats illegal UTF-8 "overlong" encodings of
    Latin-1 codepoints as _invalid_. It operates on the principle that we can
    determine whether something is Latin-1 based entirely on how long the sequence
    is expected to be.
- The move of `SetIndent` to before `StartDocument`, which the abstraction layer
  immediately undoes. This is to accommodate XML writers that require indent to be
  set _before_ the document starts.

DHowett committed:
xar: move libxml2 behind a thin abstraction layer

This commit prepares write...xar for another XML writing backend.

DHowett committed:
tests: add a xar write test for a unicode filename

DHowett commented:
Related to #1811, the fix for which builds on these changes at https://github.com/DHowett/libarchive/compare/feature/xml-abstraction...DHowett:libarchive:feature/xmllite.

DHowett commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -95,9 +97,26 @@
+static int xml_writer_destroy(struct xml_writer *ctx);
 
-#define BAD_CAST_CONST (const xmlChar *)
+/*#define DEBUG_PRINT_TOC		1 */
```
This line was moved from elsewhere.

DHowett commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -3252,4 +3290,116 @@
+
+static int
+xml_writer_create(struct xml_writer **pctx) {
+	struct xml_writer *ctx = calloc(1, sizeof(struct xml_writer));
```
(Resolved)

DHowett commented:
Same failure on macOS observed in libarchive/libarchive@master as well.

evelikov commented:
@DHowett note that I'm not a maintainer of the project so take the following with a pinch of salt.

As a person who hears xmllite for the first time, can you give us a 10 word brief - where it's available, why should we add support (performance, features, other)?

That said overall the PR looks good - just a couple of small comment incoming.

evelikov commented on libarchive/test/test_write_format_xar.c:
```diff
@@ -274,6 +306,13 @@ test_xar(const char *option)
+	/* xar mandates the use of UTF-8 XML; if we cannot
+	 * use UTF-8, perhaps we should not write xar. */
+	if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
+		skipping("en_US.UTF-8 locale not available on this system.");
```
What are the visible side-effects of not using that locale? Should we move the actual set/check locale in the library instead?

Why aren't we checking other UTF-8 locates - be that `en_gb.UTF-8`, `ko.UTF-8` ... or the recently introduced `C.UTF-8` one? The last one might be Linux and/or glibc specific.

DHowett replied:
That's a really good question! I don't think the library takes a position on whether _formats_ work if certain encodings aren't available... but it sure does take a stance on whether tests can be run!

```
test_zip_filename_encoding.c:   if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_zip_7075_utf8_paths.c: if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_ustar_filename.c:      if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_ustar_filename.c:      if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_ustar_filename.c:      if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_tar_filename.c:        if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_rar.c:  if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
test_read_format_lha_filename_utf16.c:  if (NULL == setlocale(LC_ALL, "en_US.UTF-8")) {
```

evelikov replied:
I wonder if using `export LC_ALL=....` in the test suite runners (cmake/autotools) won't be a more scalable and consistent experience?

Either way - having your comment in the commit message, is great for posterity et al.

evelikov commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -95,9 +97,26 @@ archive_write_set_format_xar(struct archive *_a)
+static int xml_writer_end_element(struct xml_writer *ctx);
+static int xml_writer_get_final_content_and_length(struct xml_writer *ctx,
+    const char **out, size_t *size);
+static int xml_writer_destroy(struct xml_writer *ctx);
```
If it were me, I would move the whole abstraction layer to a new header, effectively removing the forward declarations. Bonus point - the debug line should no longer be considered new code :-)

DHowett replied:
That's fair, and is a good call.

However -- in reading through some of the issues on this project and looking at some of the design decisions, I get the feeling that the maintainers value isolated, single-file abstractions like this. Nobody else needs an XML writer today, and putting it into a header namespace tells other developers that they can take a dependency on it ... which would fail to link if XAR wasn't being built!

> _Quote from Tim Kientzle on [PR 1599](https://github.com/libarchive/libarchive/pull/1599)_
> A side effect of the above principles is that we do naturally get some duplication of common basic functionality. Personally, I'm fine with that: it has the side effect of keeping most source files standalone and easy to understand without having to dig through the entire system to untangle a lot of specialized vernacular. It means that most library functionality could be implemented outside of the library without any problems.

I'm happy to let @mmatuska or Tim overrule my conscience and say "yes, please split it into a separate file" however! :smile:

evelikov replied:
I'm not sure how moving this abstraction to a header opposes/violates the "single-file abstractions" being mentioned?

evelikov commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -1134,74 +1152,94 @@ make_fflags_entry(struct archive_write *a, xmlTextWriterPtr writer,
-	free(tmp);
-	if (r < 0) {
-		r = xmlTextWriterStartElement(writer, BAD_CAST("name"));
+	if (!is_u8_zstring_latin1(file->basename.s)) {
```
The u8 helper should be a separate commit with the commit message containing parts you've added in the PR summary. Also `s/inefficient/doing X instead of Y/`

DHowett replied:
ACKed, will do! It is both objectively inefficient _and_ doing X instead of Y. I'll note that.

Thanks!

evelikov commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -3252,4 +3290,120 @@ getalgname(enum sumalg sumalg)
+
+static int
+xml_writer_create(struct xml_writer **pctx) {
+	struct xml_writer *ctx = calloc(1, sizeof(struct xml_writer));
```
Nits:
 - move the opening brace for functions, at the start of the next line
 - drop unused function return type - xml_writer_destroy, xml_writer_set_indent, xml_writer_get_final_content_and_length
 - s/return (x)/return x/ - the style varies across the project, the latter being my personal preference

DHowett replied:
Great catches!

- I'll move the braces. That was my mix-up, Thanks!
- Those returns are required for part 2, https://github.com/DHowett/libarchive/compare/feature/xml-abstraction...DHowett:libarchive:feature/xmllite (the Windows APIs can fail here!)
- Ah, in this case I went with the prevailing style for this _file_ (nit: I forgot to use `()` for a couple bits). Of the 203 `return`s in the document as of 097ecfb8, 100% of them are `()`-style. I'm intending to standardize on that one :smile: even though it's also not my personal style.

evelikov replied:
> Those returns are required for part 2 ...

In that case, they should be introduced as needed. ATM it's dead code that static analyzers and people will flag up.

evelikov commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -3252,4 +3290,120 @@ getalgname(enum sumalg sumalg)
+	if (ctx->writer == NULL) {
+		xmlBufferFree(ctx->bp);
+		free(ctx);
+		return (-1);
```
The caller will print a bunch of `-1` messages now on error, I'd suggest swapping these with ENOMEM.

DHowett replied:
Great call. Thanks!

DHowett replied:
So, thinking about this one more: the original code did not indicate failure with `ENOMEM` here either. Just that "it failed"!

I'm on the fence about switching these ones over, since to uphold the consistent contract that negative return values are errors it would need to be `-ENOMEM` and then we're in ever-hotter water because that's really not a useful error code _either_.
The internal consumer would dutifully report it as `-12` (or whatever)... but at the end of the day it would still be a fatal failure.

evelikov replied:
The original code returns an `ARCHIVE_ERRNO_MISC` with "xmlNewTextWriterMemory() couldn't create xml writer". The new  `ARCHIVE_ERRNO_MISC` with  "xml_writer_create() failed: -1".

Both aren't particularly useful, although since this PR is adding an abstraction it shouldn't really change it IMHO.

evelikov commented on libarchive/archive_write_set_format_xar.c:
```diff
@@ -1544,51 +1583,43 @@ make_toc(struct archive_write *a)
 		goto exit_toc;
 	}
-	r = xmlTextWriterSetIndent(writer, 4);
+	r = xml_writer_start_document(writer);
```
Keeping the SetIndent <> StartDocument swap as separate commit would be great, but not required IMHO.

mmatuska commented:
Any news on this?

DHowett commented:
> Any news on this?

I’m circling back to it 😄 

I have some misgivings about the change in UTF-8 validation behavior, and I wouldn’t want to introduce an issue that could come back to bite libarchive.

DHowett commented:
So, regarding the change in UTF-8 validation behavior.

As far as I can tell, the _worst_ outcome is that libarchive will base64-encode more Latin1 string names than it would before. I do not believe that to be an issue.

> **Note**
> It looks like `archive_entry_..._pathname_utf8` rejects the overlong encodings specified below immediately, so all of the below is academic and/or moot. I realized this after writing up the theory. 😁

---

For each case noted below, I will be referring to two archive entries `test1` and `test\xE0\x80\xB1` (which I _believe_ is an overlong encoding for `0x31`).

- Before
	- `UTF8Toisolat1` would(?) return `1` for both indicating that they were both Latin1
	- We would send both encodings over to libxml2.
	- If libxml2 accepted it, the archive would contain two files that appear to have the same name.
	- It may be encoded in the final output document with percent escapes or XML entities.
- After
	- `is_u8_zstring_latin1` would return `1` for `test1` and `0` for the overlong encoding.
	- We would base64-encode the overlong encoding.
	- The archive would contain two files that appear to have the same name.

Now, the result in both cases is the same: the recipient will probably pass that name on to a filesystem API or display it to the user. What happens after that is anybody's guess.

However... none of that matters, because you could also _just create a normal archive that contains the same filename two times_ without using UTF-8 trickery. :smile:

DHowett commented:
And just for final confirmation: The xar writer runs all filenames through `xar->sconv`, which is initialized to be a `UTF-8` string converter. It converts the overlong encoding into `\uFFFF` before the new code ever sees it.

DHowett committed:
xar: replace UTF8Toisolat1 with a local reimplementation

In xar, Non-Latin-1 gets stored Base64-encoded. However:

- `UTF8Toisolat1` is a `libxml2` export.
- The writer code made a copy of every filename for the purposes of
	checking whether it was Latin-1 stored as UTF-8.

Instead of moving it into an abstraction with the rest of the libxml2
imports, I have elected to replace it with a simple UTF-8 validator.

The new function has slightly different results than the one it is
replacing for invalid UTF-8. Namely, it treats illegal UTF-8 "overlong"
encodings of Latin-1 codepoints as _invalid_. It operates on the
principle that we can determine whether something is Latin-1 based
entirely on how long the sequence is expected to be.

In practice, this does not matter because the xar writer passes all
filenames through a UTF-8 converter that rejects overlong encodings
before they get here.

Therefore, this replacement produces the same results as UTF8Toisolat1
with one fewer import and one fewer copy per filename.

DHowett commented:
Ah, and last -- I missed @evelikov's question!

> As a person who hears xmllite for the first time, can you give us a 10 word brief - where it's available, why should we add support (performance, features, other)?

`xmllite` is an XML library that comes with all versions of Windows since (I believe) Windows XP.

We should add support for it so that Windows libarchive consumers can offer xar with a reduced footprint and without taking on any additional non-platform dependencies.

---

Selfishly, I'm the person who is currently responsible for the Windows built-in copy of libarchive and I would love to add xar support without introducing _another_ XML library to the system install directory :grin:

evelikov commented:
Neat, I had no idea `xmllite` is available so far back. I can sympathise with the "selfish" reasons you mentioned.

OOC: Does the Windows built-in copy of libarchive have any out-of-tree patches?  If so are there any plans on send them upstream - hint: happy to throw some (virtual) :cookie: your way if you do :-P

DHowett commented:
> OOC: Does the Windows built-in copy of libarchive have any out-of-tree patches? If so are there any plans on send them upstream - hint: happy to throw some (virtual) 🍪 your way if you do :-P

We have only one out-of-tree patch outstanding! https://github.com/DHowett/libarchive/commit/ed33fc3b5a8a09c72ebb2339c0fc8e53fd5796fb

The rest, I've already upstreamed:
- #1622
- #1623

The one I haven't submitted yet is mostly sitting in my outgoing queue because I am not in love with the mechanics of it. The Windows build system requires the use of the secure `s` variants of `{w,s,f}*open`, but I haven't done the legwork to figure out whether they're available on every Win32-like platform libarchive supports, so I've been hanging on to the patch until I can be sure it's good for the entire ecosystem. :smile:

mmatuska commented:
Could you rebase the patch please? We would be happy to make the library compatible with xmllite

DHowett commented:
Hey there! Is there any work remaining on my side to land this patch series? I'd love to start the review for the next phase -- adding `xmllite` support to XAR -- which depends on this one! 🙂

DHowett commented:
As far as I can tell, the remaining questions are stylistic rather than semantic -- whether the error message indicates that "xmlCreateWriter" failed to create a writer or whether "xml_writer_create" failed, and the specifics of some dead code introduced as hooks for the next pull request. I'd love to drive those down, but I don't see them as being significantly detracting from the overall pull request.

I'd be happy to either keep this as a separate PR or integrate it with the following work for introducing xmllite so it can be reviewed as a single unit.

Thoughts?

PS:

> If you're looking for a way to get some (virtual) 🍪 is to fix the Windows/Mingw build.

My team has recently contributed to the Windows Unicode correctness and we're making sure we can stand up tests for MSVC and revive the Windows Mingw builds as well. :)
